### PR TITLE
OP-1530 Packaging deb with systemd and documentation for Charlie

### PR DESCRIPTION
**Pull request by sacherjj**
Adding all needed files to debian package.

`casper-node` and `casper-client` are installed in `/usr/bin`

All other files are currently in `/etc/casper/`.  

Allows removable and reinstall of .deb while maintaining `/etc/casper/config.toml' and keys in '/etc/casper/validator_keys`.  This will allow distribution of updated `accounts.csv` and `chainspec.toml` with a .deb package install.

### CODE REVIEW

File: resources/charlie/config-example.toml [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+# ====================================
+# Configuration options for networking
+# ====================================
+[network]
+
+# The public address of the node.
+#
+# It must be publicly available in order to allow peers to connect to this node.
+# If the port is set to 0, the actual bound port will be substituted.
+public_address = '[IP ADDRESS]:0'
```
Comment (sacherjj) [inline resources/charlie/config-example.toml:46]: @Fraser999 Should this be `:0` or `:34553`?

Comment (Fraser999) [reply resources/charlie/config-example.toml:46]: Best to leave it as `0`.  The node only uses this to gossip its public address to peers so they can connect to it.  If `0` is specified, it's automatically updated to reflect the actual bound port once the node starts listening, so it's probably safer to just always leave it set to `0`.

File: resources/charlie/config-example.toml [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+gossip_interval = 30000
+
+
+# =============================================
+# Configuration options for the HTTP API server
+# =============================================
+[http_server]
+
+# Interface to bind to.  Defaults to loopback address.
+bind_interface = '127.0.0.1'
```
Comment (Fraser999) [inline resources/charlie/config-example.toml:72]: This means only a client on the same machine will be able to access the HTTP server.  Not sure if that's ok for charlie or not.  We could default to bind to `0.0.0.0` here too I guess.

File: resources/charlie/validator_keys/README.md [hunk 3]
```diff
@@ -0,0 +1,19 @@
+## Validator Keys  
+
+Keys should be created by running (assuming you are in this directory for local path "."):
```
Comment (Fraser999) [inline resources/charlie/validator_keys/README.md:3]:
```suggestion
Keys should be created by running (assuming you are in `/etc/casper/validator_keys/` for local path "."):
```

File: resources/charlie/validator_keys/README.md [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+`sudo casper-client keygen .`
+
+This will create `public_key_hex`, `public_key.pem`, and `secret_key.pem`.
```
Comment (Fraser999) [inline resources/charlie/validator_keys/README.md:7]: Probably worth a line about never distributing the `secret_key.pem` file (except maybe to Clarity)?

File: resources/charlie/README.md [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+
+The package can be installed with `sudo apt install casper-node_0.1.x-r_amd64.deb`.  
+Where `x` is last minor version of software and `r` is the package revision. We will be pushing revisions
+with no code changes for updates to genesis files.
+
+This package will install both `casper-node` and `casper-client` executables in `/usr/bin`.
+
+Configuration files and other needed files are installed in `/etc/casper/`. And example config file is given
+as `/etc/casper/config-example.toml`. This needs to be updated to `config.toml`. You can do a direct copy and 
+modify values as needed. We did not distribute `config.toml` so it isn't overwritten when a deb package is
```
Comment (Fraser999) [inline resources/charlie/README.md:15]: Do you think you need to mention about running the text editor as sudo?

File: resources/charlie/README.md [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+Unless specified in different location in `config.toml` the data for the node will
+be located in `/root/.local/share/casper-node` for a linux system.
+
+## Resetting and upgrading for new run
+
+Prior to a new upgrade and run, we need to clean up our local state.
+
+```
+sudo systemctl stop casper-node
+sudo apt remove casper-node
```
Comment (Fraser999) [inline resources/charlie/README.md:149]: When running this, I get the following warning:
```
dpkg: warning: while removing casper-node, directory '/etc/casper/validator_keys' not empty so not removed
```
Worth mentioning or not?

File: resources/charlie/validator_keys/README.md (hunk 3 above)
Comment (Fraser999) [inline resources/charlie/validator_keys/README.md:3]: I expect we won't clear out this folder when upgrading or uninstalling, so should the docs here mention that keys only need creating if they don't exist already?

File: resources/charlie/config-example.toml [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+# Configuration options for the storage component
+# ===============================================
+[storage]
+
+# Optional path (absolute, or relative to this config.toml) to the folder where any files created
+# or read by the storage component will exist.
+#
+# If the folder doesn't exist, it and any required parents will be created.
+#
+# If unset, defaults to:
```
Comment (Fraser999) [inline resources/charlie/config-example.toml:88]: These defaults should be updated to reflect that the node will run as root.  I guess the Windows entry can be removed (I was a bit optimistic including that one!)

File: resources/charlie/README.md [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+sudo rm /root/.local/share/casper-node/*
```
Comment (Fraser999) [inline resources/charlie/README.md:150]: This command didn't work for me.  I had to do
```
sudo su -
rm /root/.local/share/casper-node/*
exit
```

File: resources/charlie/CHANGE_LOG.md [hunk 9]
```diff
@@ -0,0 +1,7 @@
+# Change Log
+
+## casper-node_0.1.0-0_amd64.deb
+
+ - Initial Release
+ - Actions Required:
+    - Create '/etc/casper/config.toml' from '/etc/casper/config-example.toml'
```
Comment (Fraser999) [inline resources/charlie/CHANGE_LOG.md:7]: Maybe mention keygen too?

**Review by Fraser999 — COMMENTED** [review]
Nice work!  Some nitpicks and questions, but I could pretty much just follow the instructions as is.

File: node/Cargo.toml [hunk 10]
```diff
@@ -46,7 +46,7 @@ log = { version = "0.4.8", features = ["std", "serde", "kv_unstable"] }
 num = { version = "0.2.0", default-features = false }
 num-derive = "0.3.0"
 num-traits = "0.2.10"
-openssl = "0.10.29"
+openssl = { version = "0.10.29", features = ["vendored"] }
```
[the remark this answers is not in the document]
Comment (sacherjj) [reply node/Cargo.toml:49]: We are fighting install on Ubuntu 18.04 and this has been required by Archit.  But I can try that flag in the build for packaging.

[the remark this answers is not in the document]
Comment (Fraser999) [reply node/Cargo.toml:49]: No probs.

If we find that we're always building with that feature, then we should remove the `vendored-openssl` feature from our own Cargo.toml.

**Comment by sacherjj** [comment]
bors try

File: resources/charlie/README.md [hunk 11]
```diff
@@ -158,3 +158,7 @@ sudo apt install casper-node_0.1.x-r_amd64.deb`
  
 Look at `/etc/casper/CHANGE_LOG.md` for any changes that require action from
 validators. We might need to update `config.toml` to support different fields, for example.
+
+## Bugs
+
+Please file and bugs as issues on [the casper-node GitHub repo](https://github.com/CasperLabs/casper-node).
```
Comment (Fraser999) [inline resources/charlie/README.md:164]:
```suggestion
Please file any bugs as issues on [the casper-node GitHub repo](https://github.com/CasperLabs/casper-node).
```

File: resources/charlie/config-example.toml [hunk 12]
```diff
@@ -90,8 +90,8 @@ bind_port = 7777
 #   /home/alice/.local/share/casper-node
 # * macOS: `$HOME/Library/Application Support/io.CasperLabs.casper-node`, e.g.
 #   /Users/Alice/Library/Application Support/io.CasperLabs.casper-node
-# * Windows: `{FOLDERID_RoamingAppData}\CasperLabs\casper-node\data` e.g.
-#   C:\Users\Alice\AppData\Roaming\CasperLabs\casper-node\data
+# * With packaged systemd service, the location will be
```
Comment (Fraser999) [inline resources/charlie/config-example.toml:93]: This change leaves in the outdated info for Linux and macOS above.

File: node/Cargo.toml (hunk 10 above)
[the remark this answers is not in the document]
Comment (sacherjj) [reply node/Cargo.toml:49]: Just tested pulling this and it still works fine, so this will be gone.

File: resources/charlie/README.md (hunk 8 above)
Comment (sacherjj) [reply resources/charlie/README.md]: I'm suggesting `sudo rm -rf /root/.local/share/casper-node` as that seems to work and still be a one liner.

File: resources/charlie/config-example.toml (hunk 12 above)
Comment (sacherjj) [reply resources/charlie/config-example.toml]: Removed.

**Review by MParlikar — APPROVED** [review]

File: resources/charlie/README.md [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+installed to upgrade.
+
+In the section of the config named `[network]`, change the `<IP ADDRESS>` in `public_address` to your server's external IP.
+
+The `accounts.csv` and `chainspec.toml` files will be installed in `/etc/casper` with the deb package install. 
+This should allow easy configuration for each network run as they are updated with a new deb package revision.
+
+## Key Generation
+
+Follow commands in `/etc/casper/validator_keys/README.md` to generate keys and get public key hex to send in.
```
[the remark this answers is not in the document]
Comment (sacherjj) [reply resources/charlie/README.md:34]: Added details to the README.md in validator_keys for this.

**Pull request merged**